### ISSUE TITLE
Name of CSRF token header

### ISSUE THREAD

**Issue by chkal** [mvc-spec/mvc-spec#207]
I'm currently working on the TCK tests for the CSRF protection and noticed something strange. The spec states:

> MVC implementations are REQUIRED to support CSRF tokens [...] **as HTTP headers**.

> Any other value than CsrfOptions.OFF will automatically inject a CSRF token as an HTTP header. **The actual name of this header is implementation dependent**.

> Automatic validation is enabled by setting this property to CsrfOptions.IMPLICIT, in which case **all
post requests must include either an HTTP header** or a hidden field with the correct token. 

So to sum up: If you want to send the CSRF token with a hidden field, you simply render a HTML hidden element like this:

```html
<input type="hidden" name="${mvc.csrf.name}" value="${mvc.csrf.token}"/>
```

This means you get the name of the hidden field from the MVC `Csrf` instance.

If you want to use the HTTP header method of communicating the token, you must lookup the token from the HTTP response header (most likely on the client side in JavaScript). But the name of the header is implementation dependent. Then you have to send this token with an HTTP request header back to the server, but this name of this header is also not specified and therefore varies between implementations.

Sounds like you cannot use the HTTP header method in a portable way at all. Or am I missing something?

**Comment by chkal** [comment]
I'm done with the TCK tests for the Security chapter of the spec. For now, I just hardcoded the header name used by Krazo:

https://github.com/mvc-spec/mvc-tck/blob/master/tests/src/main/java/org/mvcspec/tck/tests/security/CsrfConstants.java#L20

I think we have a few options for this issue:

1. Leave it as it is. I don't think this is a good option.
2. We specify a fixed header name. But I don't think there is a common header name supported by all client-side frameworks. The current value used by Krazo is `X-CSRF-TOKEN` and is automatically supported by Angular. AFAIK it is also the default used by Spring Security.
3. We provide a portable way to configure the name of the header and keep the default as being implementation specific.
4. We provide a portable way to configure the name of the header and define a default in the spec (maybe we could use `X-CSRF-TOKEN`?)

@ivargrimstad @lefloh @mvc-spec/expert-group Thoughts?

**Comment by ivargrimstad** [comment]
I am leaning towards 2. If it is good enough for Spring Security, it should be good enough for us.
But I guess it would also be nice if we let it be configurable. In that case, I think the spec should specify the default value as in 4.

**Comment by lefloh** [comment]
For me it's important that the developer can change the name. I'd also be fine with a non-portable solution (Implementations MUST provide a way to configure the header name).
But option 4 is definitely the nicest one 👍

**Comment by chkal** [comment]
Option 4 is also my favorite. I guess we all like this one best. I'll prepare a pull request for the spec.

**Comment by chkal** [comment]
Fixed via #209
